HDDS-9530. Recon - NPE in handling deleteKey event in NSSummaryFSO task.

devmadhuu opened the pull request:
## What changes were proposed in this pull request?

This PR fixes by omitting the event to be handled by Recon where a value in a table could be null for a key. Ideally this should not happen in Raft log sequence of events so this PR is trying to skip such event and log with WARNING and details of event like table, event key etc. This will address 2 issues along with update to javadoc for `processEvent` method code in`OMDBUpdatesHandler.java`

1. NPE will not be thrown while processing event (PUT, UPDATE, DELETE) in Recon task handler, if a event is having null value for a key.
2. Any inconsistency or issue identified later using above log message will be deeply analyzed why and on which table such key occurrence can happen on OM DB side.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9530

## How was this patch tested?

This patch was tested using existing Junit tests as well by creating and deleting keys using freon tool and wait for Recon sync to happen with OM DB.

devmadhuu commented:
@adoroszlai @dombizita @sumitagrawl pls review.

Galsza reviewed (changes requested):
Left 2 comments suggesting changes. Could you also create a test that covers this change?

devmadhuu commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/OMDBUpdatesHandler.java:
```diff
@@ -153,8 +153,18 @@ private void processEvent(int cfIndex, byte[] keyBytes, byte[]
+          }
+          LOG.info(
+              "WARNING !!!, Value of Key: {} in table: {} should not be null " +
+                  "for DELETE event ", keyStr, tableName);
```
> I think converting the key to string and the keyStr variable is not needed. You could use key.toString() here instead. (omUpdateEventValidator.isValidEvent() does exectly this a few lines later)

Thanks @Galsza for review, as per my understanding DTOKEN_TABLE is using OzoneTokenIdentifier as key instead of String, so I believe we still need to check type of key, but yes we can simply do toString()... mostly this `isValidEvent` method has not returned false, so log.warn here not hitting else this might throw exception... What do you think ?

devmadhuu commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/OMDBUpdatesHandler.java:
```diff
@@ -153,8 +153,18 @@ private void processEvent(int cfIndex, byte[] keyBytes, byte[]
+          if (key instanceof String) {
+            keyStr = (String) key;
+          }
+          LOG.info(
```
Thanks @Galsza for review. Just didn't realize it while writing code. I'll change it to log.warn

devmadhuu commented:
> Left 2 comments suggesting changes. Could you also create a test that covers this change?

@Galsza , Thanks for review. Handled delete events test case assertions for events where null value or non existent key scenario can come. Such events will be logged and will not flown down to handle. Ideally we should have DELETE events with old null value,  so doing warn logging  and avoid flowing down.

Galsza commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/OMDBUpdatesHandler.java:
```diff
@@ -153,8 +153,14 @@ private void processEvent(int cfIndex, byte[] keyBytes, byte[]
-            oldValue, key, action)) {
+        if (null == oldValue) {
+          String keyStr = (key instanceof String) ? key.toString() : "";
+          LOG.warn("Value of Key: {} in table: {} should not be null " +
```
In case that the key is not a string, we are logging `"Value of Key: in table: 'tableName' should not be null".` Could you please improve it a bit more so the person encountering the log message understand what happened here?

devmadhuu replied:
@Galsza Thanks for re-review. I have added bit more information in case of key as non-String type. Pls check and review once.

Galsza commented:
Thank you @devmadhuu for the patch it looks good to me now. 

@dombizita and @sumitagrawl could you please take a look at this?

ArafatKhan2198 reviewed (changes requested):
Thanks for the changes @devmadhuu , since we're already tweaking the file, let's give the comments and JavaDoc a quick change too. It's not directly related to your changes, but it's good to keep everything clear and easy to understand.

1. Could you please improve the `processEvent()` Java doc to :- 

https://github.com/apache/ozone/blob/5e80d8283bd13878d55a2f3a241586930c97f32e/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/OMDBUpdatesHandler.java#L91
```
/**
 * Processes a database update event based on the provided parameters.
 *
 * @param cfIndex     Index of the column family.
 * @param keyBytes    Serialized key bytes.
 * @param valueBytes  Serialized value bytes.
 * @param action      Type of the database action (e.g., PUT, DELETE).
 * @throws IOException If an I/O error occurs.
 */
```
2. Shorten this comment to :- "**_Tag PUT operations on existing keys as "UPDATE_**" events."
https://github.com/apache/ozone/blob/5e80d8283bd13878d55a2f3a241586930c97f32e/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/OMDBUpdatesHandler.java#L140

3. Shorten this comment to :- "**_Log and ignore events if key or value types are undetermined_**" https://github.com/apache/ozone/blob/5e80d8283bd13878d55a2f3a241586930c97f32e/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/OMDBUpdatesHandler.java#L187

4. Improved the description of this comment for better understanding :- 
```
// Handle the event based on its type:
// - PUT with a new key: Insert the new value.
// - PUT with an existing key: Update the existing value.
// - DELETE with an existing key: Remove the value.
// - DELETE with a non-existing key: No action, log a warning if necessary.
```
https://github.com/apache/ozone/blob/5e80d8283bd13878d55a2f3a241586930c97f32e/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/OMDBUpdatesHandler.java#L114

ArafatKhan2198 reviewed (changes requested):
Could you change this as well to :- `if (action.equals(DELETE))`

https://github.com/apache/ozone/blob/5e80d8283bd13878d55a2f3a241586930c97f32e/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/OMDBUpdatesHandler.java#L130

ArafatKhan2198 commented:
If it's possible we could rephrase the paragraph at the starting of the process() method better clarity and understanding.
Please let me know if I have gotten anything wrong in my understanding.
```
// Note on DTOKEN_TABLE handling:
// - The DTOKEN_TABLE uniquely utilizes OzoneTokenIdentifier, not a String, as its key.
// - Directly casting to String during deserialization would result in errors.
// - The omdbLatestUpdateEvents currently maps keys as Strings. Changing this to Object would 
//   significantly affect all ReconOmTasks, hence it remains unchanged for now.
// - As DTOKEN_TABLE data is not currently required for synchronization in the Recon OM DB snapshot,
//   no immediate changes are needed.
// - Future integration of DTOKEN_TABLE data will involve storing and retrieving events using Object-type keys.
```


https://github.com/apache/ozone/blob/5e80d8283bd13878d55a2f3a241586930c97f32e/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/OMDBUpdatesHandler.java#L95

devmadhuu commented:
> // Handle the event based on its type:
> // - PUT with a new key: Insert the new value.
> // - PUT with an existing key: Update the existing value.
> // - DELETE with an existing key: Remove the value.
> // - DELETE with a non-existing key: No action, log a warning if necessary.

Thanks @ArafatKhan2198 for review and noticing that a method was left empty for its javadoc comments. May be because we follow for not adding javadoc for private methods. Nonetheless , we should add comments wherever necessary for our own understanding. Thanks again for pointing this out. I have added now and rectified all comments for easy understanding per your suggestion.

> If it's possible we could rephrase the paragraph at the starting of the process() method better clarity and understanding. Please let me know if I have gotten anything wrong in my understanding.
> 
> ```
> // Note on DTOKEN_TABLE handling:
> // - The DTOKEN_TABLE uniquely utilizes OzoneTokenIdentifier, not a String, as its key.
> // - Directly casting to String during deserialization would result in errors.
> // - The omdbLatestUpdateEvents currently maps keys as Strings. Changing this to Object would 
> //   significantly affect all ReconOmTasks, hence it remains unchanged for now.
> // - As DTOKEN_TABLE data is not currently required for synchronization in the Recon OM DB snapshot,
> //   no immediate changes are needed.
> // - Future integration of DTOKEN_TABLE data will involve storing and retrieving events using Object-type keys.
> ```
> 
> https://github.com/apache/ozone/blob/5e80d8283bd13878d55a2f3a241586930c97f32e/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/OMDBUpdatesHandler.java#L95

I read the above existing comment and it looks ok to me with more information, so leaving it as it is as of now. I also observed that this comment was modified as part of #HDDS-7455.

> Could you change this as well to :- `if (action.equals(PUT))`
> 
> https://github.com/apache/ozone/blob/5e80d8283bd13878d55a2f3a241586930c97f32e/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/OMDBUpdatesHandler.java#L130

Done.

sumitagrawl reviewed (approved):
@devmadhuu Thanks for working over this, LGTM

ArafatKhan2198 reviewed (approved):
Thanks for working on this @devmadhuu LGTM +1

dombizita reviewed (commented):
Thanks for working on this @devmadhuu, I have 2 minor comments inline, beside that it is looking good to me! Could you update just a bit the PR description that you also fixed some javadoc in this patch?

devmadhuu commented:
> Thanks for working on this @devmadhuu, I have 2 minor comments inline, beside that it is looking good to me! Could you update just a bit the PR description that you also fixed some javadoc in this patch?

Thanks @dombizita for your review. I have updated the PR description.

devmadhuu commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/OMDBUpdatesHandler.java:
```diff
@@ -153,8 +155,22 @@ private void processEvent(int cfIndex, byte[] keyBytes, byte[]
       } else if (action.equals(DELETE)) {
-        if (oldValue != null && !omUpdateEventValidator.isValidEvent(tableName,
-            oldValue, key, action)) {
+        if (null == oldValue) {
+          String keyStr = (key instanceof String) ? key.toString() : "";
+          if (keyStr.isEmpty()) {
+            LOG.warn(
+                "Only DTOKEN_TABLE table uses OzoneTokenIdentifier as key " +
+                    "instead of String. Event on any other table in this " +
+                    "condition may need to be investigated. This DELETE " +
+                    "event is on {} table which is not useful for Recon to " +
+                    "capture.", tableName);
+          }
+          LOG.warn("Old Value of Key: {} in table: {} should not be null " +
+              "for DELETE event ", keyStr, tableName);
+          return;
```
@dombizita thanks for your review, Not sure If I understood correctly, but as I can see that if key is not empty, then also we are printing that "old value is null" information using message -> "Old Value of Key: {} in table: {} should not be null " , so I believe the purpose is still getting solved. Pls let me know if I understood something not correct.

dombizita replied:
I wanted to say that with this solution I suggested we won't print it (as the `key instanceof String` will be false) and I'd like to know which approach is better: print it or not. I think it doesn't add any information with the empty key name, maybe we can adjust the second (in my suggestion) warn message to have something similar.

devmadhuu replied:
@dombizita , I feel, current approach is better because if oldValue is null, which is more important for us, so we are printing the message that "Old value of key {} in table {} should not be null" and here we are printing both non-empty key as well as table name associated with that key which has its old value as null. This should be more helpful in debugging later. First log message which is getting printed is on condition `if (keyStr.isEmpty())` and this condition also is inside the condition` if (null == oldValue)` , so key can only be empty if it is not of string type and non string type is anyway not relevant for Recon to capture, but still this log message will be helpful and see if event is really based on `DTOKEN_TABLE` with what is intended for.

dombizita replied:
Thanks for taking your time to answer here, I'm fine with this way!

devmadhuu commented on hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/tasks/TestOMDBUpdatesHandler.java:
```diff
@@ -197,18 +195,8 @@ public void testDelete() throws Exception {
-        nonExistVolume.getAction());
-    assertEquals(nonExistVolumeKey, nonExistVolume.getKey());
-    assertNull(nonExistVolume.getValue());
+    // Assert for non existent keys, no events will be captured and handled.
+    assertEquals(2, events.size());
```
Handled.

dombizita reviewed (approved):
Thanks for your continuous effort @devmadhuu, it looks good to me!
